// @module-federation/rsbuild-plugin/dist/utils/index.mjs
import util from "util";
import { DEFAULT_ASSET_PREFIX } from "../constant.mjs";
import { autoDeleteSplitChunkCacheGroups } from "./autoDeleteSplitChunkCacheGroups.mjs";
import { addDataFetchExposes } from "./addDataFetchExposes.mjs";
import { updateStatsAndManifest } from "./manifest.mjs";
import { SSR_DIR, SSR_ENV_NAME, createSSRMFConfig, createSSRREnvConfig, patchSSRRspackConfig, setSSREnv } from "./ssr.mjs";

;// CONCATENATED MODULE: external "util"

;// CONCATENATED MODULE: external "../constant.mjs"

;// CONCATENATED MODULE: external "./autoDeleteSplitChunkCacheGroups.mjs"

;// CONCATENATED MODULE: external "./addDataFetchExposes.mjs"

;// CONCATENATED MODULE: external "./manifest.mjs"

;// CONCATENATED MODULE: external "./ssr.mjs"

;// CONCATENATED MODULE: ./src/utils/index.ts

function isRegExp(target) {
    return util.types.isRegExp(target);
}






export { DEFAULT_ASSET_PREFIX, SSR_DIR, SSR_ENV_NAME, addDataFetchExposes, autoDeleteSplitChunkCacheGroups, createSSRMFConfig, createSSRREnvConfig, isRegExp, patchSSRRspackConfig, setSSREnv, updateStatsAndManifest };
