// @module-federation/rsbuild-plugin/dist/utils/ssr.mjs
import path from "path";
import { createRequire } from "node:module";
import { encodeName } from "@module-federation/sdk";
import { CALL_NAME_MAP } from "../constant.mjs";

;// CONCATENATED MODULE: external "path"

;// CONCATENATED MODULE: external "node:module"

;// CONCATENATED MODULE: external "@module-federation/sdk"

;// CONCATENATED MODULE: external "../constant.mjs"

;// CONCATENATED MODULE: ./src/utils/ssr.ts




const ssr_require = createRequire(import.meta.url);
const resolve = ssr_require.resolve;
function resolveOrRequest(request) {
    try {
        return resolve(request);
    } catch  {
        return request;
    }
}
function safeRequire(request) {
    try {
        return ssr_require(request);
    } catch  {
        return undefined;
    }
}
const SSR_DIR = 'ssr';
const SSR_ENV_NAME = 'mf-ssr';
const ENV_NAME = 'mf';
function setSSREnv() {
    process.env['MF_SSR_PRJ'] = 'true';
}
const isDev = ()=>{
    return process.env['NODE_ENV'] === 'development';
};
function patchNodeConfig(config, mfConfig) {
    var _config_output;
    config.output ||= {};
    if (config.output.publicPath === 'auto') {
        config.output.publicPath = '';
    }
    config.target = 'async-node';
    // Force node federation output to CJS + async chunk loading.
    // This prevents browser jsonp runtime handlers from leaking into SSR remotes.
    config.output.module = false;
    config.output.chunkFormat = 'commonjs';
    config.output.chunkLoading = 'async-node';
    delete config.output.chunkLoadingGlobal;
    const UniverseEntryChunkTrackerPluginModule = safeRequire('@module-federation/node/universe-entry-chunk-tracker-plugin');
    const UniverseEntryChunkTrackerPlugin = UniverseEntryChunkTrackerPluginModule === null || UniverseEntryChunkTrackerPluginModule === void 0 ? void 0 : UniverseEntryChunkTrackerPluginModule.default;
    config.plugins ||= [];
    if (isDev() && UniverseEntryChunkTrackerPlugin) {
        config.plugins.push(new UniverseEntryChunkTrackerPlugin());
    }
    const uniqueName = mfConfig.name || ((_config_output = config.output) === null || _config_output === void 0 ? void 0 : _config_output.uniqueName);
    const chunkFileName = config.output.chunkFilename;
    if (typeof chunkFileName === 'string' && uniqueName && !chunkFileName.includes(uniqueName)) {
        const encodedName = encodeName(uniqueName);
        config.output.chunkFilename = chunkFileName.endsWith('.js') ? chunkFileName.replace(/\.js$/, `${encodedName}.js`) : `${chunkFileName}${encodedName}`;
    }
}
function patchSSRRspackConfig(config, mfConfig, ssrDir, callerName, resetEntry = true, modifyPublicPath = true) {
    config.output ||= {};
    if (modifyPublicPath) {
        var _config_output;
        if (typeof ((_config_output = config.output) === null || _config_output === void 0 ? void 0 : _config_output.publicPath) !== 'string') {
            throw new Error('publicPath must be string!');
        }
        const publicPath = config.output.publicPath;
        if (publicPath !== 'auto') {
            const publicPathWithSSRDir = `${publicPath}${ssrDir}/`;
            config.output.publicPath = publicPathWithSSRDir;
        }
    }
    if (callerName === CALL_NAME_MAP.RSPRESS && resetEntry) {
        // set virtue entry, only need mf entry
        config.entry = 'data:application/node;base64,';
    }
    patchNodeConfig(config, mfConfig);
    return config;
}
function patchToolsTspack(envConfig, fn) {
    var _envConfig_tools;
    const rspackArr = [];
    if ((_envConfig_tools = envConfig.tools) === null || _envConfig_tools === void 0 ? void 0 : _envConfig_tools.rspack) {
        var _envConfig_tools1, _envConfig_tools2;
        if (Array.isArray((_envConfig_tools1 = envConfig.tools) === null || _envConfig_tools1 === void 0 ? void 0 : _envConfig_tools1.rspack)) {
            var _envConfig_tools3;
            rspackArr.push(...(_envConfig_tools3 = envConfig.tools) === null || _envConfig_tools3 === void 0 ? void 0 : _envConfig_tools3.rspack);
        } else if (typeof ((_envConfig_tools2 = envConfig.tools) === null || _envConfig_tools2 === void 0 ? void 0 : _envConfig_tools2.rspack) === 'function') {
            var _envConfig_tools4;
            rspackArr.push((_envConfig_tools4 = envConfig.tools) === null || _envConfig_tools4 === void 0 ? void 0 : _envConfig_tools4.rspack);
        }
    }
    envConfig.tools ||= {};
    envConfig.tools.rspack = [
        ...rspackArr,
        fn
    ];
}
function createSSRREnvConfig(envConfig, mfConfig, ssrDir, rsbuildConfig, callerName) {
    var _envConfig_tools, _ssrEnvConfig_output, _ssrEnvConfig_output_distPath, _ssrEnvConfig_output1, _rsbuildConfig_output_distPath, _rsbuildConfig_output;
    const rspackArr = [];
    if ((_envConfig_tools = envConfig.tools) === null || _envConfig_tools === void 0 ? void 0 : _envConfig_tools.rspack) {
        var _envConfig_tools1, _envConfig_tools2;
        if (Array.isArray((_envConfig_tools1 = envConfig.tools) === null || _envConfig_tools1 === void 0 ? void 0 : _envConfig_tools1.rspack)) {
            var _envConfig_tools3;
            rspackArr.push(...(_envConfig_tools3 = envConfig.tools) === null || _envConfig_tools3 === void 0 ? void 0 : _envConfig_tools3.rspack);
        } else if (typeof ((_envConfig_tools2 = envConfig.tools) === null || _envConfig_tools2 === void 0 ? void 0 : _envConfig_tools2.rspack) === 'function') {
            var _envConfig_tools4;
            rspackArr.push((_envConfig_tools4 = envConfig.tools) === null || _envConfig_tools4 === void 0 ? void 0 : _envConfig_tools4.rspack);
        }
    }
    const ssrEnvConfig = {
        ...envConfig
    };
    patchToolsTspack(ssrEnvConfig, (config, { environment })=>{
        if (environment.name !== SSR_ENV_NAME) {
            return;
        }
        patchSSRRspackConfig(config, mfConfig, ssrDir, callerName);
    });
    ssrEnvConfig.output = {
        ...ssrEnvConfig.output,
        // https://rsbuild.rs/config/output/target#other-targets
        // Rsbuild not support all rspack targets, so modify to async-node in modifyRspackConfig
        target: 'node',
        distPath: {
            ...(_ssrEnvConfig_output = ssrEnvConfig.output) === null || _ssrEnvConfig_output === void 0 ? void 0 : _ssrEnvConfig_output.distPath,
            root: path.join(((_ssrEnvConfig_output1 = ssrEnvConfig.output) === null || _ssrEnvConfig_output1 === void 0 ? void 0 : (_ssrEnvConfig_output_distPath = _ssrEnvConfig_output1.distPath) === null || _ssrEnvConfig_output_distPath === void 0 ? void 0 : _ssrEnvConfig_output_distPath.root) || ((_rsbuildConfig_output = rsbuildConfig.output) === null || _rsbuildConfig_output === void 0 ? void 0 : (_rsbuildConfig_output_distPath = _rsbuildConfig_output.distPath) === null || _rsbuildConfig_output_distPath === void 0 ? void 0 : _rsbuildConfig_output_distPath.root) || '', ssrDir)
        },
        emitAssets: true
    };
    return ssrEnvConfig;
}
function patchNodeMFConfig(mfConfig) {
    var _mfConfig_library;
    if (mfConfig.remotes) {
        mfConfig.remoteType = 'script';
    }
    mfConfig.exposes = {
        ...mfConfig.exposes
    };
    mfConfig.library = {
        ...mfConfig.library,
        name: mfConfig.name,
        type: ((_mfConfig_library = mfConfig.library) === null || _mfConfig_library === void 0 ? void 0 : _mfConfig_library.type) ?? 'commonjs-module'
    };
    mfConfig.runtimePlugins = [
        ...mfConfig.runtimePlugins || []
    ];
    mfConfig.runtimePlugins.push(resolveOrRequest('@module-federation/node/runtimePlugin'));
    if (isDev()) {
        mfConfig.runtimePlugins.push(resolveOrRequest('@module-federation/node/record-dynamic-remote-entry-hash-plugin'));
    }
}
function createSSRMFConfig(mfConfig) {
    const ssrMFConfig = {
        ...mfConfig
    };
    patchNodeMFConfig(ssrMFConfig);
    ssrMFConfig.dts = false;
    ssrMFConfig.dev = false;
    return ssrMFConfig;
}

export { ENV_NAME, SSR_DIR, SSR_ENV_NAME, createSSRMFConfig, createSSRREnvConfig, patchNodeConfig, patchNodeMFConfig, patchSSRRspackConfig, patchToolsTspack, setSSREnv };
